src/Features/Core/Portable/RemoveUnusedMembers/AbstractAvoidUnusedMembersCodeFixProvider.cs

```diff
@@ -0,0 +1,116 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Editing;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.AvoidUnusedMembers
+{
+    internal abstract class AbstractAvoidUnusedMembersCodeFixProvider<TFieldDeclarationSyntax> : SyntaxEditorBasedCodeFixProvider
+        where TFieldDeclarationSyntax : SyntaxNode
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds
+            => ImmutableArray.Create(IDEDiagnosticIds.AvoidUnusedMembersDiagnosticId);
+
+        // Adjust declarators to remove based on whether or not all variable declarators within a field declaration should be removed.
+        protected abstract void AdjustDeclarators(HashSet<TFieldDeclarationSyntax> fieldDeclarators, HashSet<SyntaxNode> declarators);
+
+        public override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            context.RegisterCodeFix(new MyCodeAction(
+                c => FixAsync(context.Document, context.Diagnostics[0], c)),
+                context.Diagnostics);
+            return Task.CompletedTask;
+        }
+
+        protected override Task FixAllAsync(
+            Document document,
+            ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor,
+            CancellationToken cancellationToken)
+        {
+            return FixWithEditorAsync(document, editor, diagnostics, cancellationToken);
+        }
+
+        private async Task FixWithEditorAsync(
+            Document document, SyntaxEditor editor, ImmutableArray<Diagnostic> diagnostics,
+            CancellationToken cancellationToken)
+        {
+            var declarators = new HashSet<SyntaxNode>();
+            var fieldDeclarators = new HashSet<TFieldDeclarationSyntax>();
+
+            // Compute declarators to remove, and also track common field declarators.
+            foreach (var diagnostic in diagnostics)
+            {
+                var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+                var diagnosticSpan = diagnostic.Location.SourceSpan;
+                var symbolName = diagnostic.Properties[AvoidUnusedMembersDiagnosticAnalyzer.UnunsedMemberNameProperty];
+                var symbolKind = diagnostic.Properties[AvoidUnusedMembersDiagnosticAnalyzer.UnunsedMemberKindProperty];
+
+                var node = GetTopmostSyntaxNodeForSymbolDeclaration(root.FindNode(diagnosticSpan),
+                    isSymbolDeclarationNode: n => n != null && semanticModel.GetDeclaredSymbol(n, cancellationToken)?.Name == symbolName);
```

Comment:
i'm confused why this helper function would need to be provided.  Are there other situations where you pass a different helper?  If not, why not have GetTopmostSyntaxNodeForSymbolDeclaration contain this code internally?

Reply:
That method is a virtual function, which is overridden in VB (we need to map the StatementSyntax to BlockSyntax)

---
In reply to: [212728972](https://github.com/dotnet/roslyn/pull/29511#discussion_r212728972) [](ancestors = 212728972)

Reply:
in general, we don't use that pattern.  Instead, we keep things abstract, making C# and VB have to decide what behavior they want.  It's weird to have a defualt behavior that is then only used by one language, and then have the other language override it.  In essence, this is saying "VB is overriding C#'s behavior".  And instead we want to say: it's up to each language to decide what the correct behavior is.

Also see later note: we have a service that already does this work.  So i don't think this extension point is needed in this feature.  Look up ISymbolDeclarationService :)